File: srv_erp/srv_erp/report/items_ordered_in_date_range/items_ordered_in_date_range.py
# Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import cint

from srv_erp.srv_erp.report.hierarchical_filters import get_descendant_condition
from srv_erp.srv_erp.report.uom_utils import add_selected_uom_columns


DEFAULT_BRAND_VARIANT_ATTRIBUTE = "Brand"
RESOLVED_BRAND_SQL = "COALESCE(NULLIF(variant_brand.attribute_value, ''), NULLIF(item.brand, ''), template.brand)"


def execute(filters=None):
	filters = frappe._dict(filters or {})
	set_default_warehouse(filters)
	validate_warehouse(filters)
	view = filters.get("view") or ("Item Summary" if cint(filters.get("group_by_item", 1)) else "Detailed")
	columns = get_columns(view)
	data = get_data(filters, view)
	add_selected_uom_columns(columns, data, filters.get("include_uom"))
	return columns, data


def get_columns(view="Detailed"):
	if view == "Item Summary":
		return get_grouped_columns()
	if view == "Sub-total":
		return get_subtotal_columns()

	return [
		{"label": _("Item Code"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 120},
		{"label": _("Item Name"), "fieldname": "item_name", "fieldtype": "Data", "width": 160},
		{"label": _("Item Group"), "fieldname": "item_group", "fieldtype": "Link", "options": "Item Group", "width": 120},
		{"label": _("Brand"), "fieldname": "brand", "fieldtype": "Link", "options": "Brand", "width": 120},
		{"label": _("Customer"), "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 140},
		{"label": _("Customer Name"), "fieldname": "customer_name", "fieldtype": "Data", "width": 160},
		{"label": _("Customer Group"), "fieldname": "customer_group", "fieldtype": "Link", "options": "Customer Group", "width": 130},
		{"label": _("Territory"), "fieldname": "territory", "fieldtype": "Link", "options": "Territory", "width": 120},
		{"label": _("Sales Person"), "fieldname": "sales_person", "fieldtype": "Data", "width": 140},
		{"label": _("Order Date"), "fieldname": "transaction_date", "fieldtype": "Date", "width": 100},
		{"label": _("Delivery Date"), "fieldname": "delivery_date", "fieldtype": "Date", "width": 100},
		{"label": _("Qty Ordered"), "fieldname": "qty", "fieldtype": "Float", "width": 105},
		{"label": _("Qty Ordered UOM"), "fieldname": "uom_qty", "fieldtype": "Link", "options": "UOM", "width": 105},
		{"label": _("Stock Available"), "fieldname": "stock_available_qty", "fieldtype": "Float", "width": 115, "convertible": "qty"},
		{"label": _("Stock Available UOM"), "fieldname": "uom_stock_available_qty", "fieldtype": "Link", "options": "UOM", "width": 125},
		{"label": _("Difference (Stock - Ordered)"), "fieldname": "difference_qty", "fieldtype": "Float", "width": 150},
		{"label": _("Difference Stock UOM"), "fieldname": "uom_difference_qty", "fieldtype": "Link", "options": "UOM", "width": 125},
		{"label": _("Stock Qty Delivered"), "fieldname": "stock_delivered_qty", "fieldtype": "Float", "width": 130, "convertible": "qty"},
		{"label": _("Stock Qty Delivered UOM"), "fieldname": "uom_stock_delivered_qty", "fieldtype": "Link", "options": "UOM", "width": 130},
		{"label": _("Stock Qty Pending"), "fieldname": "stock_pending_qty", "fieldtype": "Float", "width": 120, "convertible": "qty"},
		{"label": _("Stock Qty Pending UOM"), "fieldname": "uom_stock_pending_qty", "fieldtype": "Link", "options": "UOM", "width": 125},
		{"label": _("Sales Order"), "fieldname": "sales_order", "fieldtype": "Link", "options": "Sales Order", "width": 130},
		{"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 110},
		{"label": _("Project"), "fieldname": "project", "fieldtype": "Link", "options": "Project", "width": 120},
		{"label": _("Company"), "fieldname": "company", "fieldtype": "Link", "options": "Company", "width": 120},
	]


def get_subtotal_columns():
	return [
		{"label": _("Sales Order"), "fieldname": "sales_order", "fieldtype": "Link", "options": "Sales Order", "width": 140},
		{"label": _("Item Code"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 130},
		{"label": _("Item Name"), "fieldname": "item_name", "fieldtype": "Data", "width": 200},
		{"label": _("Brand"), "fieldname": "brand", "fieldtype": "Link", "options": "Brand", "width": 120},
		{"label": _("Qty Ordered"), "fieldname": "qty", "fieldtype": "Float", "width": 115},
		{"label": _("UOM"), "fieldname": "uom_qty", "fieldtype": "Link", "options": "UOM", "width": 90},
		{"label": _("Customer"), "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 150},
		{"label": _("Order Date"), "fieldname": "transaction_date", "fieldtype": "Date", "width": 105},
		{"label": _("Company"), "fieldname": "company", "fieldtype": "Link", "options": "Company", "width": 120},
	]


def get_grouped_columns():
	return [
		{"label": _("Brand"), "fieldname": "brand", "fieldtype": "Link", "options": "Brand", "width": 120},
		{"label": _("Item Code"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 120},
		{"label": _("Item Name"), "fieldname": "item_name", "fieldtype": "Data", "width": 180},
		{"label": _("Item Group"), "fieldname": "item_group", "fieldtype": "Link", "options": "Item Group", "width": 120},
		{"label": _("Qty Ordered"), "fieldname": "qty", "fieldtype": "Float", "width": 105},
		{"label": _("Qty Ordered UOM"), "fieldname": "uom_qty", "fieldtype": "Link", "options": "UOM", "width": 105},
		{"label": _("Stock Available"), "fieldname": "stock_available_qty", "fieldtype": "Float", "width": 115, "convertible": "qty"},
		{"label": _("Stock Available UOM"), "fieldname": "uom_stock_available_qty", "fieldtype": "Link", "options": "UOM", "width": 125},
		{"label": _("Difference (Stock - Ordered)"), "fieldname": "difference_qty", "fieldtype": "Float", "width": 150},
		{"label": _("Difference Stock UOM"), "fieldname": "uom_difference_qty", "fieldtype": "Link", "options": "UOM", "width": 125},
		{"label": _("Stock Qty Delivered"), "fieldname": "stock_delivered_qty", "fieldtype": "Float", "width": 130, "convertible": "qty"},
		{"label": _("Stock Qty Delivered UOM"), "fieldname": "uom_stock_delivered_qty", "fieldtype": "Link", "options": "UOM", "width": 130},
		{"label": _("Stock Qty Pending"), "fieldname": "stock_pending_qty", "fieldtype": "Float", "width": 120, "convertible": "qty"},
		{"label": _("Stock Qty Pending UOM"), "fieldname": "uom_stock_pending_qty", "fieldtype": "Link", "options": "UOM", "width": 125},
		{"label": _("Order Count"), "fieldname": "order_count", "fieldtype": "Int", "width": 95},
		{"label": _("Company"), "fieldname": "company", "fieldtype": "Link", "options": "Company", "width": 120},
	]


def get_data(filters, view="Detailed"):
	filters["brand_variant_attribute"] = (
		frappe.db.get_single_value("SRV Settings", "variant_auto_create_attribute")
		or DEFAULT_BRAND_VARIANT_ATTRIBUTE
	)
	conditions = get_conditions(filters)
	if view == "Item Summary":
		return get_grouped_data(filters, conditions)
	if view == "Sub-total":
		return get_subtotal_data(filters, conditions)

	return frappe.db.sql(
		f"""
			SELECT
				soi.item_code,
				soi.item_name,
				soi.item_group,
				{RESOLVED_BRAND_SQL} AS brand,
				so.customer,
				so.customer_name,
				so.customer_group,
				so.territory,
				(
					SELECT GROUP_CONCAT(DISTINCT st.sales_person ORDER BY st.idx SEPARATOR ', ')
					FROM `tabSales Team` st
					WHERE st.parent = so.name AND st.parenttype = 'Sales Order'
				) AS sales_person,
				so.transaction_date,
				soi.delivery_date,
				soi.uom AS uom_qty,
				soi.qty,
				COALESCE(bin.actual_qty, 0) AS stock_available_qty,
				item.stock_uom AS uom_stock_available_qty,
				COALESCE(bin.actual_qty, 0) - soi.stock_qty AS difference_qty,
				item.stock_uom AS uom_difference_qty,
				soi.delivered_qty AS stock_delivered_qty,
				soi.stock_uom AS uom_stock_delivered_qty,
				GREATEST(soi.stock_qty - soi.delivered_qty, 0) AS stock_pending_qty,
				soi.stock_uom AS uom_stock_pending_qty,
				so.name AS sales_order,
				so.status,
				so.project,
				so.company
			FROM `tabSales Order Item` soi
			INNER JOIN `tabSales Order` so ON so.name = soi.parent
			INNER JOIN `tabItem` item ON item.name = soi.item_code
			LEFT JOIN `tabBin` bin
				ON bin.item_code = soi.item_code AND bin.warehouse = %(warehouse)s
			LEFT JOIN `tabItem` template ON template.name = item.variant_of
			LEFT JOIN `tabItem Variant Attribute` variant_brand
				ON variant_brand.parent = item.name
				AND variant_brand.attribute = %(brand_variant_attribute)s
			WHERE so.docstatus = 1 {conditions}
			ORDER BY
				({RESOLVED_BRAND_SQL}) IS NULL, {RESOLVED_BRAND_SQL},
				soi.item_name, soi.item_code, so.transaction_date DESC, so.name, soi.idx
		""",
		filters,
		as_dict=True,
	)


def get_subtotal_data(filters, conditions):
	rows = frappe.db.sql(
		f"""
			SELECT
				so.name AS sales_order,
				soi.item_code,
				soi.item_name,
				{RESOLVED_BRAND_SQL} AS brand,
				SUM(soi.qty) AS qty,
				soi.uom AS uom_qty,
				so.customer,
				so.transaction_date,
				so.company
			FROM `tabSales Order Item` soi
			INNER JOIN `tabSales Order` so ON so.name = soi.parent
			INNER JOIN `tabItem` item ON item.name = soi.item_code
			LEFT JOIN `tabItem` template ON template.name = item.variant_of
			LEFT JOIN `tabItem Variant Attribute` variant_brand
				ON variant_brand.parent = item.name
				AND variant_brand.attribute = %(brand_variant_attribute)s
			WHERE so.docstatus = 1 {conditions}
			GROUP BY
				so.name, soi.item_code, soi.item_name, {RESOLVED_BRAND_SQL},
				soi.uom, so.customer, so.transaction_date, so.company
			ORDER BY so.transaction_date DESC, so.name, MIN(soi.idx)
		""",
		filters,
		as_dict=True,
	)

	return append_sales_order_subtotals(rows)


def append_sales_order_subtotals(rows):
	data = []
	current_order = None
	uom_totals = {}

	for row in rows:
		if current_order and row.sales_order != current_order:
			data.extend(make_subtotal_rows(current_order, uom_totals))
			uom_totals = {}
		current_order = row.sales_order
		uom_totals[row.uom_qty] = uom_totals.get(row.uom_qty, 0) + (row.qty or 0)
		data.append(row)

	if current_order:
		data.extend(make_subtotal_rows(current_order, uom_totals))

	return data


def make_subtotal_rows(sales_order, uom_totals):
	return [
		frappe._dict(
			{
				"sales_order": sales_order,
				"item_name": _("Sub-total"),
				"qty": qty,
				"uom_qty": uom,
				"bold": 1,
			}
		)
		for uom, qty in uom_totals.items()
	]


def get_grouped_data(filters, conditions):
	return frappe.db.sql(
		f"""
			SELECT
				{RESOLVED_BRAND_SQL} AS brand,
				soi.item_code,
				soi.item_name,
				soi.item_group,
				soi.uom AS uom_qty,
				SUM(soi.qty) AS qty,
				MAX(COALESCE(bin.actual_qty, 0)) AS stock_available_qty,
				item.stock_uom AS uom_stock_available_qty,
				MAX(COALESCE(bin.actual_qty, 0)) - SUM(soi.stock_qty) AS difference_qty,
				item.stock_uom AS uom_difference_qty,
				SUM(soi.delivered_qty) AS stock_delivered_qty,
				soi.stock_uom AS uom_stock_delivered_qty,
				SUM(GREATEST(soi.stock_qty - soi.delivered_qty, 0)) AS stock_pending_qty,
				soi.stock_uom AS uom_stock_pending_qty,
				COUNT(DISTINCT so.name) AS order_count,
				so.company
			FROM `tabSales Order Item` soi
			INNER JOIN `tabSales Order` so ON so.name = soi.parent
			INNER JOIN `tabItem` item ON item.name = soi.item_code
			LEFT JOIN `tabBin` bin
				ON bin.item_code = soi.item_code AND bin.warehouse = %(warehouse)s
			LEFT JOIN `tabItem` template ON template.name = item.variant_of
			LEFT JOIN `tabItem Variant Attribute` variant_brand
				ON variant_brand.parent = item.name
				AND variant_brand.attribute = %(brand_variant_attribute)s
			WHERE so.docstatus = 1 {conditions}
			GROUP BY
				{RESOLVED_BRAND_SQL}, soi.item_code, soi.item_name, soi.item_group,
				soi.uom, soi.stock_uom, item.stock_uom, so.company
			ORDER BY
				({RESOLVED_BRAND_SQL}) IS NULL, {RESOLVED_BRAND_SQL},
				soi.item_name, soi.item_code, soi.uom
		""",
		filters,
		as_dict=True,
	)


def get_conditions(filters):
	conditions = []
	field_map = {
		"company": "so.company",
		"customer": "so.customer",
		"territory": "so.territory",
		"project": "so.project",
		"item_code": "soi.item_code",
	}

	if filters.get("from_date"):
		conditions.append("so.transaction_date >= %(from_date)s")
	if filters.get("to_date"):
		conditions.append("so.transaction_date <= %(to_date)s")
	for fieldname, column in field_map.items():
		if filters.get(fieldname):
			conditions.append(f"{column} = %({fieldname})s")
	if filters.get("customer_group"):
		conditions.append(
			get_descendant_condition("Customer Group", "so.customer_group", "customer_group")
		)
	if filters.get("item_group"):
		conditions.append(get_descendant_condition("Item Group", "soi.item_group", "item_group"))
	if filters.get("brand"):
		conditions.append(f"{RESOLVED_BRAND_SQL} = %(brand)s")
	if filters.get("sales_person"):
		conditions.append(
			"EXISTS (SELECT 1 FROM `tabSales Team` st_filter "
			"WHERE st_filter.parent = so.name AND st_filter.parenttype = 'Sales Order' "
			"AND st_filter.sales_person = %(sales_person)s)"
		)

	return " AND " + " AND ".join(conditions) if conditions else ""


def set_default_warehouse(filters):
	if filters.get("warehouse") or not filters.get("company"):
		return

	filters["warehouse"] = frappe.db.get_value(
		"Warehouse",
		{
			"company": filters.company,
			"warehouse_name": "Finished Goods",
			"is_group": 0,
			"disabled": 0,
		},
		"name",
	)

def validate_warehouse(filters):
	if not filters.get("warehouse"):
		frappe.throw(
			_("Please select a Warehouse. No enabled Finished Goods warehouse was found for {0}.").format(
				filters.company
			)
		)

	warehouse = frappe.db.get_value(
		"Warehouse",
		filters.warehouse,
		["company", "is_group", "disabled"],
		as_dict=True,
	)
	if not warehouse or warehouse.disabled or warehouse.is_group:
		frappe.throw(_("Please select an enabled, non-group Warehouse."))
	if warehouse.company != filters.company:
		frappe.throw(_("Warehouse {0} does not belong to company {1}.").format(filters.warehouse, filters.company))

File: srv_erp/srv_erp/report/items_ordered_in_date_range/items_ordered_in_date_range.js
// Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Items Ordered in Date Range"] = {
	onload: async function(report) {
		await set_finished_warehouse(report);
	},
	filters: [
		{
			fieldname: "group_by_item",
			label: __("Group by Item"),
			fieldtype: "Check",
			default: 1,
		},
		{
			fieldname: "subtotal_view",
			label: __("Sub-total View"),
			fieldtype: "Check",
			default: 0,
			description: __("Planning quantities show their actual UOM beside every value."),
		},
		{
			fieldname: "pending_only",
			label: __("Pending Only"),
			fieldtype: "Check",
			default: 0,
			description: __("Show only Sales Order items with quantity still pending for delivery."),
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			reqd: 1,
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		{
			fieldname: "range",
			label: __("Range"),
			fieldtype: "Select",
			options: ["Daily", "Weekly", "Monthly", "Quarterly", "Yearly"],
			default: "Monthly",
			on_change() {
				const range = frappe.query_report.get_filter_value("range");
				const today = frappe.datetime.get_today();
				const offsets = { Daily: 0, Weekly: -7, Monthly: -1, Quarterly: -3, Yearly: -12 };
				const from_date = range === "Weekly"
					? frappe.datetime.add_days(today, offsets[range])
					: frappe.datetime.add_months(today, offsets[range]);
				frappe.query_report.set_filter_value("from_date", from_date);
				frappe.query_report.set_filter_value("to_date", today);
			},
		},
		{ fieldname: "customer", label: __("Customer"), fieldtype: "Link", options: "Customer" },
		{ fieldname: "customer_group", label: __("Customer Group"), fieldtype: "Link", options: "Customer Group" },
		{ fieldname: "territory", label: __("Territory"), fieldtype: "Link", options: "Territory" },
		{ fieldname: "sales_person", label: __("Sales Person"), fieldtype: "Link", options: "Sales Person" },
		{ fieldname: "item_code", label: __("Item"), fieldtype: "Link", options: "Item" },
		{ fieldname: "item_group", label: __("Item Group"), fieldtype: "Link", options: "Item Group" },
		{ fieldname: "brand", label: __("Brand"), fieldtype: "Link", options: "Brand" },
		{
			fieldname: "include_uom",
			label: __("Preferred UOM"),
			fieldtype: "Link",
			options: "UOM",
			description: __(
				"Sub-total View converts items with a valid factor and safely keeps other items in Stock UOM."
			),
		},
		{ fieldname: "project", label: __("Project"), fieldtype: "Link", options: "Project" },
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1,
			on_change: async function() {
				await set_finished_warehouse(frappe.query_report, true);
			},
		},
		{
			fieldname: "warehouse",
			label: __("Warehouse"),
			fieldtype: "Link",
			options: "Warehouse",
			reqd: 1,
			get_query() {
				return {
					filters: {
						company: frappe.query_report.get_filter_value("company"),
						is_group: 0,
						disabled: 0,
					},
				};
			},
		},
	],
	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (!data) {
			return value;
		}
		const quantity_fields = [
			"qty",
			"stock_available_qty",
			"stock_delivered_qty",
			"stock_pending_qty",
			"stock_shortfall_qty",
		];
		if (
			frappe.query_report.get_filter_value("subtotal_view") &&
			quantity_fields.includes(column.fieldname) &&
			data[column.fieldname] != null &&
			data.stock_uom
		) {
			value = `${value} <span class="text-muted">${frappe.utils.escape_html(data.stock_uom)}</span>`;
		}

		if (data.is_subtotal_detail && column.fieldname === "brand") {
			value = `&nbsp;&nbsp;&nbsp;&nbsp;${value}`;
		}
		if (data.is_group || data.is_total) {
			value = $("<span>").html(value).css("font-weight", "bold").prop("outerHTML");
		}
		value = format_production_status(value, column, data);

		return value;
	},
	get_pdf_format(report, custom_format) {
		return report.get_filter_value("subtotal_view") ? custom_format : null;
	},
};

function format_production_status(value, column, data) {
	const subtotal_view = frappe.query_report.get_filter_value("subtotal_view");
	const ordered_field = subtotal_view ? "qty" : "stock_ordered_qty";
	const status_fields = [ordered_field, "stock_delivered_qty", "stock_available_qty", "stock_shortfall_qty"];
	const shortage = flt(data.stock_shortfall_qty);
	const ordered = flt(data[ordered_field]);
	const delivered = flt(data.stock_delivered_qty);
	const stock = flt(data.stock_available_qty);

	if (column.fieldname === "item_name" || (subtotal_view && column.fieldname === "brand")) {
		const color = shortage > 0 ? "#b3261e" : "#137333";
		return `<span style="border-left:3px solid ${color};padding-left:6px">${value}</span>`;
	}
	if (!status_fields.includes(column.fieldname)) {
		return value;
	}

	let foreground = "#1a73e8";
	let background = "#e8f0fe";
	if (column.fieldname === "stock_shortfall_qty") {
		foreground = shortage > 0 ? "#b3261e" : "#137333";
		background = shortage > 0 ? "#fce8e6" : "#e6f4ea";
	} else if (column.fieldname === "stock_available_qty") {
		const stock_is_sufficient = stock >= Math.max(ordered - delivered, 0);
		foreground = stock_is_sufficient ? "#137333" : "#b3261e";
		background = stock_is_sufficient ? "#e6f4ea" : "#fce8e6";
	} else if (column.fieldname === "stock_delivered_qty") {
		foreground = delivered > 0 ? "#137333" : "#b3261e";
		background = delivered > 0 ? "#e6f4ea" : "#fce8e6";
	}

	return `<span style="display:block;margin:-2px -4px;padding:2px 4px;border-radius:3px;color:${foreground};background:${background};font-weight:700">${value}</span>`;
}

async function set_finished_warehouse(report, replace_existing = false) {
	const warehouse_filter = report.get_filter("warehouse");
	if (!warehouse_filter || (!replace_existing && warehouse_filter.get_value())) {
		return;
	}

	const company = report.get_filter_value("company");
	if (!company) {
		return;
	}

	const warehouses = await frappe.db.get_list("Warehouse", {
		filters: {
			company,
			warehouse_name: "Finished Goods",
			is_group: 0,
			disabled: 0,
		},
		fields: ["name", "warehouse_name"],
		limit: 1,
	});
	const warehouse = warehouses[0];
	await warehouse_filter.set_value(warehouse ? warehouse.name : "");
}
